library/Scripting/Duktape/Raw.hs

```diff
@@ -144,12 +176,34 @@ foreign import capi safe "duktape.h duk_push_context_dump"
 
 -------------------------------------------------------------------------------------------------------
 
-createHeap ∷ FunPtr DukAllocFunction → FunPtr DukReallocFunction → FunPtr DukFreeFunction → Ptr () → FunPtr DukFatalFunction → IO (Maybe DuktapeCtx)
+createHeap ∷ FunPtr DukAllocFunction → FunPtr DukReallocFunction → FunPtr DukFreeFunction → InternalUData → FunPtr DukFatalFunction → IO (Maybe DuktapeCtx)
 createHeap allocf reallocf freef udata fatalf = do
-  ptr ← c_duk_create_heap allocf reallocf freef udata fatalf
+  ptr ← c_duk_create_heap allocf reallocf freef (getInternalUData udata) fatalf
   if ptr /= nullPtr
      then newForeignPtr ptr (c_duk_destroy_heap ptr) >>= newMVar >>= return . Just
      else return Nothing
 
 createHeapF ∷ FunPtr DukFatalFunction → IO (Maybe DuktapeCtx)
-createHeapF = createHeap nullFunPtr nullFunPtr nullFunPtr nullPtr
+createHeapF = createHeap nullFunPtr nullFunPtr nullFunPtr (InternalUData nullPtr)
+
+-- | A TimeoutCheck is an IO action that returns True when the current script evaluation
+-- should timeout (interpreter throws RangeError).
+createGovernedHeap ∷ FunPtr DukAllocFunction → FunPtr DukReallocFunction → FunPtr DukFreeFunction → TimeoutCheck → FunPtr DukFatalFunction → IO (Maybe DuktapeCtx)
+createGovernedHeap allocf reallocf freef timeoutCheck fatalf = do
+  (udata, release) ← wrapTimeoutCheckUData timeoutCheck
+  mctx ← createHeap allocf reallocf freef udata fatalf
+  case mctx of
+    Just ctx → withMVar ctx $ \fptr → do
+      addForeignPtrFinalizer fptr release
+      return mctx
+    Nothing → return Nothing
+  where
+  -- TimeoutCheck is wrapped to pass as void* udata in `createHeap` and will be provided (by duktape)
+  -- back to `execTimeoutCheck` when the interpreter invokes that callback.
+  wrapTimeoutCheckUData ∷ TimeoutCheck → IO (InternalUData, IO ())
+  wrapTimeoutCheckUData check = do
```

Comment:
This code is not safe from async exceptions.

Reply:
I'm not trying to dismiss this as I'm sure masking is a good idea in general, but for motivation can you give an example of an async exception we could expect and recover from? Or would this be the user of the library throwing one at us while they are building a heap for some reason?

Reply:
If the user of the lib uses `System.Timeout.timeout` or `Control.Concurrent.Async.race`, for example, the gap between any of your IO actions are potential exceptions. 

A system under heavy load will timeout more easily. I've seen this with e.g. sandboxing [tryhaskell.org](https://tryhaskell.org/). 

In the race case, if you do e.g. `race evalThis readFromCache` - i.e. return whichever completes first, that might throw in your initialisation process. 

On a long running service this will at best slowly accumulate allocated bytes. At worst you're looking at increased fragmentation because these are pinned mallocs that can't be moved around by the GC.

I'm doing research for running PureScript's JS output for server-side validation/server-side rendering so I'm evaluating whether to use hs-duktape, write my own binding that will address all my concerns from the ground up, or call out to Nodejs in a pipe and avoid the concerns altogether (and the extra work). :thinking:

Reply:
My two-cents new projects should consider using Fabrice Bellard's [quickjs](https://bellard.org/quickjs/) instead though I'm not aware of an existing Haskell binding. I doubt duk will ever have es-6 compatibility. I ended up using quickjs via a Rust binding.

Reply:
I looked at quickjs, but I don't think it has (1) the maturity required, or (2) long-term viability until someone takes over maintenance. Bellard is great and all but he tends to make things and move onto something else, leaving someone else to maintain. Whereas duktape has a proper community, documentation, a list of projects using it, an official GitHub project with an issue tracker where I can report bugs, open PRs, etc. and the expectation that it'll be maintained in five years. Small JS implementations are already niche.

YMMV.